src/soup-to-3d/index.ts

```diff
@@ -130,35 +130,47 @@ export const createBoardGeomFromSoup = (soup: AnySoupElement[]): Geom3[] => {
 
   for (const { route: mixedRoute } of traces) {
     if (mixedRoute.length < 2) continue
-    const subRoutes = mixedRoute.reduce(
-      (c, p) => {
-        // @ts-ignore
-        const lastLayer = c.current?.[c.current.length - 1]?.layer
-        if (
-          p.route_type === "via" ||
-          (p.route_type === "wire" && p.layer !== lastLayer)
-        ) {
-          if (c.current.length > 2) {
-            c.allPrev.push(c.current)
-          }
-          c.current = p.route_type === "wire" ? [p] : []
-          return c
-        }
-        c.current.push(p)
-        return c
-      },
-      {
-        current: [] as typeof mixedRoute,
-        allPrev: [] as Array<typeof mixedRoute>,
-      },
-    )
-    for (const route of subRoutes.allPrev.concat([subRoutes.current])) {
-      // TODO break into segments based on layers
+
+    // Group routes by continuous segments
+    const routeSegments: (typeof mixedRoute)[] = []
+    let currentSegment: typeof mixedRoute = [mixedRoute[0]!]
+    let currentLayer =
+      mixedRoute[0]!.route_type === "wire" ? mixedRoute[0]!.layer : "top"
+
+    for (let i = 1; i < mixedRoute.length; i++) {
+      const point = mixedRoute[i]!
+
+      if (point.route_type === "via") {
+        // Complete current segment and start a new one
+        routeSegments.push(currentSegment)
+        currentSegment = [point]
+      } else if (point.route_type === "wire" && point.layer !== currentLayer) {
+        // Complete current segment and start a new one on a different layer
+        routeSegments.push(currentSegment)
+        currentLayer = point.layer
+        currentSegment = [point]
+      } else {
+        currentSegment.push(point)
+      }
+    }
+
+    // Add the last segment
+    routeSegments.push(currentSegment)
+
+    // Render each route segment
+    for (const route of routeSegments) {
+      if (route.length < 2) continue
+
       const linePath = line(route.map((p) => [p.x, p.y]))
+      const layerSign =
+        route[0]!.route_type === "via"
+          ? route[0]!.to_layer === "top"
+            ? 1
+            : -1
+          : route[0]!.layer === "top"
+            ? 1
+            : -1
```

Comment:
so this is good, but technically there are also "inner traces" that don't fall on the top or bottom layer, there may be a more future-proof way to do this

Reply:
I will try to repro inner traces later and handle them.